src/Take.Elephant.Kafka/PartitionedKafkaReceiver.cs

```diff
@@ -0,0 +1,177 @@
+        {
+            return StartConsumerTaskIfNotAsync(cancellationToken);
+        }
+
+        public virtual Task CloseAsync(CancellationToken cancellationToken)
+        {
+            if (!_closed)
+            {
+                _partitionedConsumer.Close();
+                _closed = true;
+            }
+
+            return _consumerTask;
+        }
+
+        public event EventHandler<ExceptionEventArgs> ConsumerFailed;
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (!_closed)
+                {
+                    _partitionedConsumer.Close();
+                }
+
+                _partitionedConsumer.Dispose();
+                _cts.Dispose();
+            }
+        }
+
+        private async Task StartConsumerTaskIfNotAsync(CancellationToken cancellationToken)
+        {
+            if (_consumerTask != null)
+                return;
+
+            await _consumerStartSemaphore.WaitAsync(cancellationToken);
+            try
+            {
+                if (_consumerTask == null)
+                {
+                    _consumerTask = Task
+                        .Factory
+                        .StartNew(
+                            () => ConsumeAsync(_cts.Token),
+                            TaskCreationOptions.LongRunning)
+                        .Unwrap();
+                }
+            }
+            finally
+            {
+                _consumerStartSemaphore.Release();
+            }
+        }
+
+        private async Task ConsumeAsync(CancellationToken cancellationToken)
+        {
+            if (_partitionedConsumer.Subscription.All(s => s != Topic))
+            {
+                _partitionedConsumer.Subscribe(Topic);
+            }
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    var result = _partitionedConsumer.Consume(cancellationToken);
+                    await _channel.Writer.WriteAsync(result.Message, cancellationToken);
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    break;
+                }
+                catch (Exception ex)
+                {
+                    var handler = ConsumerFailed;
+                    if (handler != null)
+                    {
+                        handler.Invoke(this, new ExceptionEventArgs(ex));
+                    }
+                    else
+                    {
+                        Trace.TraceError("An unhandled exception occurred on KafkaReceiverQueue: {0}", ex);
+                    }
+                }
+            }
+
+            _partitionedConsumer.Unsubscribe();
+            _channel.Writer.Complete();
+        }
+
+        public void Dispose()
+        {
+            _partitionedConsumer.Close();
+            _partitionedConsumer.Dispose();
```

Comment:
These two lines are not needed here, as they already are present on `Dispose(bool)` (and that's where they should be anyway following the Dispose pattern).

Besides, you're calling `Close` unconditionally here, in contrast to elsewhere, where you check for `if (!_closed)` first.

Reply:
I agree, I will remove the unnecessary lines